File: application/app/Services/ExpenseService.php
<?php

namespace App\Services;

use App\Enums\Expenses\ExpenseSearchColumns;
use App\Enums\Expenses\ExpenseSortOptions;
use App\Enums\Expenses\ExpenseTableActions;
use App\Models\Expense;
use App\Models\ExpensePayment;
use DateInterval;
use DateTime;

class ExpenseService
{
    private Expense $expenses;
    private ExpensePayment $payments;

    public function __construct(Expense $expenses, ExpensePayment $payments)
    {
        $this->expenses = $expenses;
        $this->payments = $payments;
    }

    public function createExpense(array $expenseData): void
    {
        $createData = [
            'name' => $expenseData['name'],
            'description' => $expenseData['description'],
            'categoryId' => $expenseData['categoryId'],
            'cost' => $expenseData['cost'],
            'recurrenceRate' => $expenseData['recurrenceRate'],
            'userId' => $expenseData['userId'],
            'startDate' => $expenseData['startDate'],
            'endDate' => $expenseData['endDate'],
            'paidOnCreation' => $expenseData['paidOnCreation'],
            'initialDatePaid' => $expenseData['initialDatePaid'],
            'dueLastDayOfMonth' => $expenseData['dueLastDayOfMonth'],
        ];

        if ($expenseData['dueLastDayOfMonth']) {
            $createData['startDate'] = substr($expenseData['startDate'], 0, 8) . (new DateTime())->format('t');
        }

        $expenseId = $this->expenses->createExpense($createData);
        if ($expenseId == 0) {
            throw new \Exception("Failed to create expense");
        }

        if ($expenseData['paidOnCreation']) {
            $paidData = [
                'expenseId' => $expenseId,
                'isPaid' => true,
                'dueDate' => $expenseData['startDate'],
                'userId' => $expenseData['userId'],
                'datePaid' => $expenseData['initialDatePaid'],
            ];

            $this->updateExpensePaidStatus($paidData);
        }
    }

    public function getAllExpensesByUserId(array $requestData): array
    {
        $sortEnum = ExpenseSortOptions::fromName($requestData['sort']);
        if ($sortEnum === null) {
            throw new \Exception('Invalid Sort Option');
        }

        $data = [
            'userId' => $requestData['userId'],
            'sort' => $sortEnum->column(),
            'sortDir' => $requestData['sortDir'],
            'showInactiveExpenses' => $requestData['showInactiveExpenses']
        ];

        if ($requestData['searchColumn'] !== null) {
            $searchColumn = ExpenseSearchColumns::fromName($requestData['searchColumn']);
            if ($searchColumn === null) {
                throw new \Exception('Invalid Search Column');
            }

            $data['searchColumn'] = $searchColumn->column();
            $data['searchValue'] = $requestData['searchValue'];
        }

        $expenses = $this->expenses->getAllExpensesByUserId($data);
        foreach ($expenses as &$expense) {
            $expense['tableActions'] = [
                ExpenseTableActions::Pay->name => ExpenseTableActions::Pay->value,
                ExpenseTableActions::Unpay->name => ExpenseTableActions::Unpay->value,
                ExpenseTableActions::Delete->name => ExpenseTableActions::Delete->value,
                ExpenseTableActions::Edit->name => ExpenseTableActions::Edit->value,
                ExpenseTableActions::EditPayments->name => ExpenseTableActions::EditPayments->value,
            ];

            $expense['active'] === 1 ? $expense['tableActions'][ExpenseTableActions::Inactive->name] = ExpenseTableActions::Inactive->value
                                     : $expense['tableActions'][ExpenseTableActions::Active->name] = ExpenseTableActions::Active->value;

            unset($expense);
        }

        return $expenses;
    }

    public function getExpensesForDashboardCalendar(array $requestData): array
    {
        $year = $requestData['year'];
        $month = $requestData['month'];

        $firstDate = new DateTime($year . '-' . $month . '-01');
        $daysInMonth = $firstDate->format('t');

        $data = [
            'userId' => $requestData['userId'],
            'firstDate' => $year . '-' . $month . '-01',
            'lastDate' => $year . '-' . $month . '-' . $daysInMonth,
        ];

        $expenses = $this->expenses->getExpensesForDashboard($data);

        $mappedExpenses = [];
        for ($i = 1; $i <= $daysInMonth; $i++) {
            $date = $year . '-' . str_pad($month, 2, '0', STR_PAD_LEFT) . '-' . str_pad($i, 2, '0', STR_PAD_LEFT);
            $dateObj = new DateTime($date);
            $mappedExpenses[$date] = [];
            foreach ($expenses as $expense) {
                if ($this->expenseIsForDate($expense, $dateObj)) {
                    $mappedExpenses[$date][] = $expense;
                }
            }
        }

        return $mappedExpenses;
    }

    public function getUpcomingExpenses($userId): array
    {
        $startDateObj = new DateTime();
        $startDateYear = (int)$startDateObj->format('Y');
        $startDateMonth = (int)$startDateObj->format('m');
        $startDateDay = (int)$startDateObj->format('j');
        $daysInCurrentMonth = (int)$startDateObj->format('t');

        $endDateYear = $startDateYear;
        $endDateMonth = $startDateObj->format('m');
        $endDateDay = $startDateDay + 7;
        // List overflows into following month
        if ($endDateDay > $daysInCurrentMonth) {
            $endDateDay -= $daysInCurrentMonth;
            $endDateMonth += 1;
        }
        // List overflows into following year
        if ($endDateMonth === 13) {
            $endDateYear += 1;
            $endDateMonth = 1;
        }

        $data = [
            'userId' => $userId,
            'firstDate' => $startDateObj->format('Y-m-d'),
            'lastDate' => "{$endDateYear}-{$endDateMonth}-{$endDateDay}",
        ];

        $expenses = $this->expenses->getExpensesForDashboard($data);

        $mappedExpenses = [];
        for ($i = $startDateDay; $i < $startDateDay + 7; $i++) {
            $year = $startDateYear;
            $month = $startDateMonth;
            $day = $i;
            if ($daysInCurrentMonth < $day) {
                $month += 1;
                $day -= $daysInCurrentMonth;
            }

            if ($month === 13) {
                $year += 1;
                $month = 1;
            }

            $date = $year . '-' . str_pad($month, 2, '0', STR_PAD_LEFT) . '-' . str_pad($day, 2, '0', STR_PAD_LEFT);
            $dateObj = new DateTime($date);
            $paymentsForDate = $this->getAllPaymentsForDate(['dueDatePaid' => $date, 'userId' => $userId]);
            $mappedExpenses[$date] = [];
            foreach ($expenses as $expense) {
                if (!$this->expenseIsForDate($expense, $dateObj)) {
                    continue;
                }

                $paymentsMap = array_column($paymentsForDate, 'due_date_paid', 'expense_id');
                if (isset($paymentsMap[$expense['id']])) {
                    $expense['due_date_paid'] = $paymentsMap[$expense['id']];
                }

                $mappedExpenses[$date][] = $expense;
            }
        }

        return $mappedExpenses;
    }

    public function expenseIsForDate(array $expense, DateTime $dateObj): bool
    {
        $startDateObj = new DateTime($expense['start_date']);
        $endDateObj = $expense['end_date'] ? new DateTime($expense['end_date']) : null;
        if ($startDateObj > $dateObj || (isset($endDateObj) && $endDateObj < $dateObj)) {
            return false;
        }

        $diffDays = $dateObj->diff($startDateObj)->days;
        switch ($expense['recurrence_rate']) {
            case 'once':
                if (substr($expense['start_date'], 0, 10) == $dateObj->format('Y-m-d')) {
                    return true;
                }
                break;
            case 'weekly':
                if ($diffDays % 7 === 0) {
                    return true;
                }
                break;
            case 'monthly':
                if ($expense['due_end_of_month'] && $dateObj->format('j') == $dateObj->format('t')) {
                    return true;
                }
                if ($dateObj->format('j') == $startDateObj->format('j')) {
                    return true;
                }
                break;
            case 'yearly':
                if ($dateObj->format('m') === $startDateObj->format('m') && $dateObj->format('j') === $startDateObj->format('j')) {
                    return true;
                }
                break;
            default:
                return true;
        }

        return false;
    }

    public function deleteExpense(int $expenseId, int $userId): bool
    {
        return $this->expenses->deleteExpense($expenseId, $userId);
    }

    public function updateExpensePaidStatus(array $requestData): void
    {
        $expenseId = $requestData['expenseId'];
        $isPaid = $requestData['isPaid'];
        $dueDate = $requestData['dueDate'];
        $userId = $requestData['userId'];
        $datePaid = $requestData['datePaid'];

        $expense = $this->expenses->getExpenseById($expenseId, $userId);
        if (!$expense) {
            throw new \Exception("Expense not found");
        }

        if ($isPaid) {
            $paymentExists = $this->payments->getExpensePaymentForDueDate([
                'userId' => $userId,
                'expenseId' => $expenseId,
                'dueDate' => $dueDate
            ]);
            if ($paymentExists) {
                // TODO: Create error response and implement here
                return;
            }

            $paymentData = [
                'expenseId' => $expense['id'],
                'userId' => $expense['user_id'],
                'cost' => $expense['cost'],
                'dueDatePaid' => $dueDate,
                'paymentDate' => $datePaid !== null ? $datePaid : (new Datetime())->format('Y-m-d'),
            ];

            $this->payments->createPayment($paymentData);
        } else {
            $paymentData = [
                'expenseId' => $expense['id'],
                'dueDate' => $dueDate,
                'userId' => $expense['user_id'],
            ];

            $this->payments->deletePaymentForDueDate($paymentData);
        }

        if ($dueDate == (new DateTime($expense['next_due_date']))->format('Y-m-d')) {
            $this->updateNextDueDate($expense, $isPaid);
        }
    }

    private function updateNextDueDate(array $expense, bool $isFuture): void
    {
        $intervalMap = [
            'daily' => '1D',
            'weekly' => '1W',
            'monthly' => '1M',
            'yearly' => '1Y',
        ];

        $rate = $expense['recurrence_rate'];
        $interval = $intervalMap[$rate] ?? null;
        if (!$interval) {
            $this->expenses->updateExpense(['active' => 0], $expense['id'], $expense['user_id']);
            return;
        }
        $dateInterval = new DateInterval("P$interval");


        $paymentExistsForDueDate = true;
        $currDueDate = new DateTime($expense['next_due_date']);
        while ($paymentExistsForDueDate) {
            if ($expense['recurrence_rate'] === 'monthly' && $expense['due_end_of_month']) {
                $isFuture ? $currDueDate->modify('first day of next month')->modify('last day of this month')
                          : $currDueDate->modify('first day of this month')->modify('last day of previous month');
            } else {
                $isFuture ? $currDueDate->add($dateInterval)
                          : $currDueDate->sub($dateInterval);
            }

            if (isset($expense['end_date']) && $currDueDate->format('Y-m-d') > $expense['end_date']) {
                $this->expenses->updateExpense(['active' => 0], $expense['id'], $expense['user_id']);
                return;
            }

            $paymentExistsForDueDate = $this->payments->getExpensePaymentForDueDate([
                'expenseId' => $expense['id'],
                'dueDate' => $currDueDate->format('Y-m-d'),
                'userId' => $expense['user_id']
            ]);
        }

        $this->expenses->updateExpense(['next_due_date' => $currDueDate->format('Y-m-d')], $expense['id'], $expense['user_id']);
    }

    private function getAllPaymentsForDate(array $requestData): array
    {
        return $this->payments->getPaymentsForDate($requestData);
    }

    // TODO: Get late expenses by checking if payment for date exists
    public function getLateExpenses(int $userId): array
    {
        return $this->expenses->getLateExpenses($userId);
    }

    public function getSortOptions(): array
    {
        $sortOptions = ExpenseSortOptions::cases();
        $optionsMap = [];
        foreach ($sortOptions as $sortOption) {
            $optionsMap[$sortOption->name] = $sortOption->value;
        }

        return $optionsMap;
    }

    public function getSearchableColumns(): array
    {
        // Keep expected order for frontend
        return [
            ExpenseSearchColumns::CreatedDate->name => ExpenseSearchColumns::CreatedDate->value,
            ExpenseSearchColumns::UpdatedDate->name => ExpenseSearchColumns::UpdatedDate->value,
            ExpenseSearchColumns::Category->name => ExpenseSearchColumns::Category->value,
            ExpenseSearchColumns::Name->name => ExpenseSearchColumns::Name->value,
            ExpenseSearchColumns::Cost->name => ExpenseSearchColumns::Cost->value,
            ExpenseSearchColumns::NextDueDate->name => ExpenseSearchColumns::NextDueDate->value,
            ExpenseSearchColumns::RecurrenceRate->name => ExpenseSearchColumns::RecurrenceRate->value,
            ExpenseSearchColumns::StartDate->name => ExpenseSearchColumns::StartDate->value,
            ExpenseSearchColumns::EndDate->name => ExpenseSearchColumns::EndDate->value,
        ];
    }

    public function getExpenseTableActions(): array
    {
        $actions = ExpenseTableActions::cases();
        $actionsMap = [];
        foreach ($actions as $action) {
            $actionsMap[] = $action->name;
        }

        return $actionsMap;
    }

    public function updateActiveStatus(array $requestData): bool
    {
        return $this->expenses->updateExpense(['active' => (int)$requestData['isActive']], $requestData['expenseId'], $requestData['userId']);
    }

    public function getPaymentsForExpense($data): array {
        return $this->payments->getPaymentsForExpenseId($data);
    }
}

File: server/BuilderServices/ExpenseService/ExpenseService.cs
using AuthenticationServices;
using BuilderRepositories;
using BuilderServices.ExpenseService.Enums;
using BuilderServices.ExpenseService.Requests;
using DatabaseServices.Models;
using MySql.Data.MySqlClient;

namespace BuilderServices.ExpenseService;

public class ExpenseService
{
    private readonly ExpenseRepository _expenseRepo;

    private readonly ExpensePaymentRepository _paymentRepo;

    private readonly ScheduledPaymentRepository _scheduledPaymentRepo;

    private readonly UserContext _userContext;

    private readonly int _upcomingDaysLimit = 7;

    public ExpenseService(ExpenseRepository expenseRepo, ExpensePaymentRepository paymentRepo, UserContext userContext, ScheduledPaymentRepository scheduledPaymentRepo) 
    {
        _expenseRepo = expenseRepo;
        _paymentRepo = paymentRepo;
        _userContext = userContext;
        _scheduledPaymentRepo = scheduledPaymentRepo;
    }

    public async Task<long> CreateExpenseAsync(CreateExpenseRequest request)
    {
        var expenseDto = new ExpenseDto
        {
            Name = request.Name,
            Cost = request.Cost,
            Description = request.Description,
            RecurrenceRate = request.RecurrenceRate,
            StartDate = request.StartDate,
            EndDate = request.EndDate,
            DueEndOfMonth = request.EndOfTheMonth,
            CategoryId = request.CategoryId,
            AutomaticPayments = request.IsAutomaticPayment,
            AutomaticPaymentCreditCardId = request.AutomaticPaymentCreditCardId
        };

        if (request.EndOfTheMonth)
        {
            var year = int.Parse(expenseDto.StartDate[..4]);
            var month = int.Parse(expenseDto.StartDate.Substring(4, 2));
            expenseDto.StartDate = $"{year}-{month}-{DateTime.DaysInMonth(year, month)}";
        }

        var currentDueDate = DateOnly.ParseExact(expenseDto.StartDate, "yyyy-MM-dd");
        var today = DateOnly.FromDateTime(DateTime.Today);
        while (currentDueDate < today)
        {
            currentDueDate = expenseDto.RecurrenceRate switch
            {
                "daily" => currentDueDate.AddDays(1),

                "weekly" => currentDueDate.AddDays(7),

                "monthly" => currentDueDate.AddMonths(1),

                "yearly" => currentDueDate.AddYears(1),

                _ => currentDueDate,
            };
        }

        expenseDto.NextDueDate = currentDueDate.ToString("yyyy-MM-dd");
        
        var lastInsertedId = await _expenseRepo.CreateExpenseAsync(expenseDto, _userContext.UserId).ConfigureAwait(false);
        if (lastInsertedId == 0)
        {
            throw new GenericException("Failed to create expense");
        }

        if (request.IsAutomaticPayment)
            await _scheduledPaymentRepo.SchedulePaymentAsync((int)lastInsertedId, expenseDto.NextDueDate, request.AutomaticPaymentCreditCardId).ConfigureAwait(false);

        return lastInsertedId;
    }

    public async Task<Dictionary<string, List<ExpenseDto>>> GetExpensesForDashboardCalendarAsync(int year, int month)
    {
        var daysInMonth = DateTime.DaysInMonth(year, month);
        var firstDate = new DateOnly(year, month, 1);
        var lastDate = new DateOnly(year, month, daysInMonth);

        var expenses = await _expenseRepo.GetExpensesForDateRangeAsync(_userContext.UserId, firstDate, lastDate).ConfigureAwait(false);

        var mappedExpenses = new Dictionary<string, List<ExpenseDto>>();
        for (int i = 1; i <= daysInMonth; i++)
        {
            var date = new DateOnly(year, month, i);
            mappedExpenses[date.ToString("yyyy-MM-dd")] = [];
            foreach (var expense in expenses)
            {
                if (BuilderUtils.ExpenseIsForDate(expense, date))
                {
                    mappedExpenses[date.ToString("yyyy-MM-dd")].Add(expense);
                }
            }
        }

        return mappedExpenses;
    }

    public async Task<Dictionary<string, List<ExpenseDto>>> GetUpcomingExpensesAsync()
    {
        var startDate = DateOnly.FromDateTime(DateTime.Today);
        var endDate = startDate.AddDays(_upcomingDaysLimit);

        var expenses = await _expenseRepo.GetExpensesForDateRangeAsync(
            _userContext.UserId, 
            startDate, 
            endDate
        ).ConfigureAwait(false);

        var mappedExpenses = new Dictionary<string, List<ExpenseDto>>();
        for (DateOnly date = startDate; date < endDate; date = date.AddDays(1))
        {
            var paymentsForDate = await _paymentRepo.GetPaymentsForDateAsync(date, _userContext.UserId).ConfigureAwait(false);
            mappedExpenses[date.ToString("yyyy-MM-dd")] = [.. expenses.Where(e => BuilderUtils.ExpenseIsForDate(e, date) && !paymentsForDate.Any(p => p.ExpenseId == e.Id))];
        }

        return mappedExpenses;
    }

    public async Task<List<ExpenseDto>> GetAllExpensesForTableAsync(string sortColumn, string sortDir, string? searchColumn, string? searchValue, bool showInactive)
    {
        var expenses = await _expenseRepo.GetAllExpensesForTableAsync(_userContext.UserId, sortColumn, sortDir, searchColumn, searchValue, showInactive).ConfigureAwait(false);
        foreach (var expense in expenses)
        {
            var recurrenceIsOnce = expense.RecurrenceRate == "once";
            expense.TableActions = new Dictionary<string, string>();
            if (expense.Active)
            {
                expense.TableActions[ExpenseTableAction.Inactive.ToString()] = ExpenseTableAction.Inactive.GetActionText(recurrenceIsOnce);
                expense.TableActions[ExpenseTableAction.Pay.ToString()] = ExpenseTableAction.Pay.GetActionText(recurrenceIsOnce);
            }
            else
            {
                expense.TableActions[ExpenseTableAction.Active.ToString()] = ExpenseTableAction.Active.GetActionText(recurrenceIsOnce);
            }
            expense.TableActions[ExpenseTableAction.Unpay.ToString()] = ExpenseTableAction.Unpay.GetActionText(recurrenceIsOnce);
            expense.TableActions[ExpenseTableAction.Edit.ToString()] = ExpenseTableAction.Edit.GetActionText(recurrenceIsOnce);
            expense.TableActions[ExpenseTableAction.Delete.ToString()] = ExpenseTableAction.Delete.GetActionText(recurrenceIsOnce);

            if (recurrenceIsOnce)
            {
                var paymentExists = (await _paymentRepo.GetPaymentsForExpenseAsync(expense.Id).ConfigureAwait(false)).Count > 0;
                if (paymentExists)
                    expense.TableActions.Remove(ExpenseTableAction.Pay.ToString());
                else
                    expense.TableActions.Remove(ExpenseTableAction.Unpay.ToString());
            }
        }

        return expenses;
    }

    public static Dictionary<string, string> GetSortOptions()
    {
        var sortOptions = new Dictionary<string, string>();
        foreach (var option in Enum.GetValues<ExpenseSortOption>())
        {
            sortOptions[option.ToString()] = option.GetDisplayName();
        }

        return sortOptions;
    }

    public static Dictionary<string, string> GetSearchColumns()
    {
        // Keep expected order for frontend
        return new Dictionary<string, string>()
        {
            { ExpenseSearchColumn.CreatedDate.ToString(), ExpenseSearchColumn.CreatedDate.GetDisplayName() },
            { ExpenseSearchColumn.UpdatedDate.ToString(), ExpenseSearchColumn.UpdatedDate.GetDisplayName() },
            { ExpenseSearchColumn.Category.ToString(), ExpenseSearchColumn.Category.GetDisplayName() },
            { ExpenseSearchColumn.Name.ToString(), ExpenseSearchColumn.Name.GetDisplayName() },
            { ExpenseSearchColumn.Cost.ToString(), ExpenseSearchColumn.Cost.GetDisplayName() },
            { ExpenseSearchColumn.NextDueDate.ToString(), ExpenseSearchColumn.NextDueDate.GetDisplayName() },
            { ExpenseSearchColumn.RecurrenceRate.ToString(), ExpenseSearchColumn.RecurrenceRate.GetDisplayName() },
            { ExpenseSearchColumn.StartDate.ToString(), ExpenseSearchColumn.StartDate.GetDisplayName() },
            { ExpenseSearchColumn.EndDate.ToString(), ExpenseSearchColumn.EndDate.GetDisplayName() },
        };
    }

    public async Task UpdateExpenseAsync(int expenseId, string? name = null, double? cost = null,
        string? startDate = null, string? endDate = null, int? categoryId = null, string? description = null, int? active = null)
    {
        if (active != null && active != 1 && active != 0)
            throw new GenericException("Invalid value for active field");
        
        var values = new
        {
            name,
            cost,
            start_date = startDate,
            end_date = endDate,
            category_id = categoryId,
            description,
            active
        };

        var updateDict = new Dictionary<string, object?>();
        foreach (var prop in values.GetType().GetProperties())
        {
            var value = prop.GetValue(values);
            if (value != null)
            {
                updateDict[prop.Name] = value;
            }
        }

        await _expenseRepo.UpdateExpenseAsync(updateDict, expenseId, _userContext.UserId);
    }

    public async Task DeleteExpenseAsync(int expenseId)
    {
        await _expenseRepo.DeleteExpenseAsync(expenseId, _userContext.UserId).ConfigureAwait(false);
    }

    public async Task<List<ExpenseDto>> GetLateExpensesAsync()
    {
        var expenses = await _expenseRepo.GetAllExpensesAsync(_userContext.UserId);

        var lateExpenses = new List<ExpenseDto>();
        foreach (var expense in expenses)
        {
            var hasLate = false;
            var currentDueDate = DateOnly.ParseExact(expense.StartDate, "yyyy-MM-dd");
            var today = DateOnly.FromDateTime(DateTime.UtcNow);
            while (currentDueDate < today)
            {
                if (expense.EndDate != null && currentDueDate > DateOnly.ParseExact(expense.EndDate, "yyyy-MM-dd"))
                    break;
                
                var paymentExists = await _paymentRepo.GetExpensePaymentByDueDateAsync(currentDueDate.ToString("yyyy-MM-dd"), expense.Id) != null;
                if (!paymentExists)
                    hasLate = true;
                
                currentDueDate = expense.RecurrenceRate switch
                {
                    "daily" => currentDueDate.AddDays(1),

                    "weekly" => currentDueDate.AddDays(7),

                    "monthly" => currentDueDate.AddMonths(1),

                    "yearly" => currentDueDate.AddYears(1),

                    _ => currentDueDate,
                };
            }

            if (hasLate)
                lateExpenses.Add(expense);
        }

        return lateExpenses;
    }

    public async Task<List<string>> GetLateDatesForExpense(int expenseId)
    {
        List<string> lateDates = [];

        var expense = await _expenseRepo.GetExpenseByIdAsync(expenseId, _userContext.UserId).ConfigureAwait(false)
            ?? throw new GenericException("Failed to find expense.");

        DateOnly? endDate = expense.EndDate != null ? DateOnly.ParseExact(expense.EndDate, "yyyy-MM-dd") : null;
        var currentDate = DateOnly.ParseExact(expense.StartDate, "yyyy-MM-dd");
        var today = DateOnly.FromDateTime(DateTime.Today);
        while(currentDate < today && (endDate == null || currentDate <= endDate))
        {
            var paymentExists = (await _paymentRepo.GetExpensePaymentByDueDateAsync(currentDate.ToString("yyyy-MM-dd"), expenseId).ConfigureAwait(false)) != null;
            if (!paymentExists)
                lateDates.Add(currentDate.ToString("yyyy-MM-dd"));

            currentDate = expense.RecurrenceRate switch
            {
                "daily" => currentDate.AddDays(1),

                "weekly" => currentDate.AddDays(7),

                "monthly" => currentDate.AddMonths(1),

                "yearly" => currentDate.AddYears(1),

                _ => currentDate,
            };
        }

        return lateDates;
    }

    public static Dictionary<string, string> GetExpenseTableBatchActions()
    {
        var actions = new Dictionary<string, string>()
        {
            { ExpenseTableBatchAction.UpdateCategory.ToString(), ExpenseTableBatchAction.UpdateCategory.GetDisplayName() }
        };

        return actions;
    }
}  